src/ty_check/unify.rs

```diff
@@ -103,19 +103,6 @@ impl ConstrainMap {
     /// Compose two substitutions `other` \circ `self`
     pub fn composition(&mut self, other: ConstrainMap) {
         if !other.is_empty() {
-            self.ty_unifier
-                .iter()
-                .for_each(|(name, _)| assert!(other.ty_unifier.get(name).is_none()));
-            self.dty_unifier
-                .iter()
-                .for_each(|(name, _)| assert!(other.dty_unifier.get(name).is_none()));
-            self.nat_unifier
-                .iter()
-                .for_each(|(name, _)| assert!(other.nat_unifier.get(name).is_none()));
-            self.mem_unifier
-                .iter()
-                .for_each(|(name, _)| assert!(other.mem_unifier.get(name).is_none()));
-
```

Comment:
This assumption is removed because it was wrong.
This assumes every implicit kinded identifier is inferred (replaced by a concrete type) at most once.
This is not the case if a expression is checked multiple times like the body of a while-loop. In this case a type can be inferred multiple times.